llvm/unittests/Analysis/IR2VecTest.cpp

```diff
@@ -0,0 +1,254 @@
+  EXPECT_NE(ErrMsg.find("Unknown IR2VecKind"), std::string::npos);
+}
+
+TEST_F(IR2VecTest, AddVectors) {
+  Embedding E1 = {1.0, 2.0, 3.0};
+  Embedding E2 = {0.5, 1.5, -1.0};
+
+  TestableEmbedder::addVectors(E1, E2);
+  EXPECT_DOUBLE_EQ(E1[0], 1.5);
+  EXPECT_DOUBLE_EQ(E1[1], 3.5);
+  EXPECT_DOUBLE_EQ(E1[2], 2.0);
+
+  // Check that E2 is unchanged
+  EXPECT_DOUBLE_EQ(E2[0], 0.5);
+  EXPECT_DOUBLE_EQ(E2[1], 1.5);
+  EXPECT_DOUBLE_EQ(E2[2], -1.0);
+}
+
+TEST_F(IR2VecTest, AddScaledVector) {
+  Embedding E1 = {1.0, 2.0, 3.0};
+  Embedding E2 = {2.0, 0.5, -1.0};
+
+  TestableEmbedder::addScaledVector(E1, E2, 0.5f);
+  EXPECT_DOUBLE_EQ(E1[0], 2.0);
+  EXPECT_DOUBLE_EQ(E1[1], 2.25);
+  EXPECT_DOUBLE_EQ(E1[2], 2.5);
+
+  // Check that E2 is unchanged
+  EXPECT_DOUBLE_EQ(E2[0], 2.0);
+  EXPECT_DOUBLE_EQ(E2[1], 0.5);
+  EXPECT_DOUBLE_EQ(E2[2], -1.0);
+}
+
+#if GTEST_HAS_DEATH_TEST
+TEST_F(IR2VecTest, MismatchedDimensionsAddVectors) {
+  Embedding E1 = {1.0, 2.0};
+  Embedding E2 = {1.0};
+  EXPECT_DEATH(TestableEmbedder::addVectors(E1, E2),
+               "Vectors must have the same dimension");
+}
+
+TEST_F(IR2VecTest, MismatchedDimensionsAddScaledVector) {
+  Embedding E1 = {1.0, 2.0};
+  Embedding E2 = {1.0};
+  EXPECT_DEATH(TestableEmbedder::addScaledVector(E1, E2, 1.0f),
+               "Vectors must have the same dimension");
+}
+#endif
+
+TEST_F(IR2VecTest, LookupVocab) {
+  Vocab V = {{"foo", {1.0, 2.0}}, {"bar", {3.0, 4.0}}};
+  LLVMContext Ctx;
+  Module M("M", Ctx);
+  FunctionType *FTy = FunctionType::get(Type::getVoidTy(Ctx), false);
+  Function *F = Function::Create(FTy, Function::ExternalLinkage, "f", M);
+
+  TestableEmbedder E(*F, V, 2);
+  auto V_foo = E.lookupVocab("foo");
+  EXPECT_EQ(V_foo.size(), 2u);
+  EXPECT_DOUBLE_EQ(V_foo[0], 1.0);
+  EXPECT_DOUBLE_EQ(V_foo[1], 2.0);
+
+  auto V_missing = E.lookupVocab("missing");
+  EXPECT_EQ(V_missing.size(), 2u);
+  EXPECT_DOUBLE_EQ(V_missing[0], 0.0);
+  EXPECT_DOUBLE_EQ(V_missing[1], 0.0);
+}
+
+TEST_F(IR2VecTest, ZeroDimensionEmbedding) {
+  Embedding E1;
+  Embedding E2;
+  // Should be no-op, but not crash
+  TestableEmbedder::addVectors(E1, E2);
+  TestableEmbedder::addScaledVector(E1, E2, 1.0f);
+  EXPECT_TRUE(E1.empty());
+}
+
+TEST_F(IR2VecTest, IR2VecVocabResultValidity) {
+  // Default constructed is invalid
+  IR2VecVocabResult invalidResult;
+  EXPECT_FALSE(invalidResult.isValid());
+#if GTEST_HAS_DEATH_TEST
+  EXPECT_DEATH(invalidResult.getVocabulary(), "IR2Vec Vocabulary is invalid");
+  EXPECT_DEATH(invalidResult.getDimension(), "IR2Vec Vocabulary is invalid");
+#endif
+
+  // Valid vocab
+  Vocab V = {{"foo", {1.0, 2.0}}, {"bar", {3.0, 4.0}}};
+  IR2VecVocabResult validResult(std::move(V));
+  EXPECT_TRUE(validResult.isValid());
+  EXPECT_EQ(validResult.getDimension(), 2u);
+}
+
+// Helper to create a minimal function and embedder for getter tests
+struct GetterTestEnv {
+  Vocab V;
+  LLVMContext Ctx;
+  std::unique_ptr<Module> M;
+  Function *F;
```

Comment:
nit: init primitive types at declaration (or make them `const` to force compile time error if uninitialized by ctor). It helps avoid undefined behavior due to later maintenance.

Reply:
Done